File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static class Point{
        int x, y;
        
        Point(int x, int y){
            this.x = x;
            this.y = y;
        }
        
    }
    
    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
    
    public int solution(int[][] maps) {
        int answer = -1;
        
        int N = maps.length;
        int M = maps[0].length;
        int[][] visited = new int[N][M];
        Queue<Point> q = new LinkedList<>();
        q.add(new Point(0, 0));
        visited[0][0] = 1;
        
        while(!q.isEmpty()){
            Point now = q.poll();
            if(now.x == N-1 && now.y == M-1) break;
            for(int i = 0; i < 4; i++){
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                
                if(0 > nx || nx >= N || 0 > ny || ny >= M){
                    continue;
                }
                if(visited[nx][ny] != 0) continue;
                if(maps[nx][ny] != 1) continue;
                q.add(new Point(nx, ny));
                visited[nx][ny] = visited[now.x][now.y] + 1;
            }
        }
        return visited[N-1][M-1] != 0 ? visited[N-1][M-1] : -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    const N = maps.length;
    const M = maps[0].length;
    const visited = Array.from({length : N}, () => Array(M).fill(false));
    
    const queue = [];
    queue.push([0, 0, 1]);
    visited[0][0] = true;
    
    while(queue.length > 0){
        const [x, y, dis] = queue.shift();
        
        if(x === N-1 && y === M-1) return dis;
        
        for(let i = 0; i < 4; i++){
            const nx = x + dx[i];
            const ny = y + dy[i];
            
            if(nx < 0 || nx >= N || ny < 0 || ny >= M){
                continue;
            }if(maps[nx][ny] == 0 || visited[nx][ny] === true){
                continue;
            }
            queue.push([nx, ny, dis + 1]);
            visited[nx][ny] = true;
        }
    }
    
    return -1;
}